vowpalwabbit/slim/include/vw_slim_predict.h

```diff
@@ -0,0 +1,434 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+#include <string>
+
+// avoid mmap dependency
+#define DISABLE_SHARED_WEIGHTS
+
+#include "example_predict.h"
+#include "explore.h"
+#include "gd_predict.h"
+#include "model_parser.h"
+#include "opts.h"
+
+namespace vw_slim {
+
+	/**
+	* @brief Exploration algorithm specified by the model.
+	*/
+	enum vw_predict_exploration
+	{
+		epsilon_greedy,
+		softmax,
+		bag
+	};
+
+	uint64_t ceil_log_2(uint64_t v);
+
+	// this guard assumes that namespaces are added in order
+	// the complete feature_space of the added namespace is cleared afterwards
+	class namespace_copy_guard
+	{
+		example_predict& _ex;
+		unsigned char _ns;
+		bool _remove_ns;
+	public:
+		namespace_copy_guard(example_predict& ex, unsigned char ns);
+		~namespace_copy_guard();
+
+		void feature_push_back(feature_value v, feature_index idx);
+	};
+
+
+	class feature_offset_guard
+	{
+		example_predict& _ex;
+		uint64_t _old_ft_offset;
+	public:
+		feature_offset_guard(example_predict& ex, uint64_t ft_offset);
+		~feature_offset_guard();
+	};
+
+	class stride_shift_guard
+	{
+		example_predict& _ex;
+		uint64_t _shift;
+	public:
+		stride_shift_guard(example_predict& ex, uint64_t shift);
+		~stride_shift_guard();
+	};
+
+	/**
+	* @brief Vowpal Wabbit slim predictor. Supports: regression, multi-class classification and contextual bandits.
+	*/
+	template<typename W>
+	class vw_predict
+	{
+		std::unique_ptr<W> _weights;
+		std::string _id;
+		std::string _version;
+		std::string _command_line_arguments;
+		std::vector<std::string> _interactions;
+		bool _ignore_linear[256];
+		bool _no_constant;
+
+		vw_predict_exploration _exploration;
+		float _minimum_epsilon;
+		float _epsilon;
+		float _lambda;
+		int _bag_size;
+
+		uint32_t _stride_shift;
+		bool _model_loaded;
+
+	public:
+		vw_predict()
+			: _model_loaded(false)
+		{ }
+
+		/**
+		* @brief Reads the Vowpal Wabbit model from the supplied buffer (produced using vw -f <modelname>)
+		*
+		* @param model The binary model.
+		* @param length The length of the binary model.
+		* @return int Returns 0 (S_VW_PREDICT_OK) if succesful, otherwise one of the error codes (see E_VW_PREDICT_ERR_*).
+		*/
+		int load(const char* model, size_t length)
+		{
+			if (!model || length == 0)
+				return E_VW_PREDICT_ERR_INVALID_MODEL;
+
+			_model_loaded = false;
+
+			// required for inline_predict
+			memset(_ignore_linear, false, sizeof(_ignore_linear));
```

Comment:
This should be done in the constructor no?

Reply:
load() can be called again with a new buffer.

Reply:
When using std::array, fill can be used.

Reply:
Performance of memset() vs std::fill() ?

---
In reply to: [304044273](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1800#discussion_r304044273) [](ancestors = 304044273)